### ISSUE TITLE
Table Extracts are running extremely slow

### ISSUE THREAD

**Issue by GoogleCodeExporter** [runt18/google-bigquery#133]
```
What version of the product are you using? On what operating system?

I am using the web interface via google chrome on a mac.

Please provide any additional information below.

Table extract jobs are taking an extremely long time.  The extract job is 
running for over 6 hours.  It should run in ten minutes.

The jobid is savvy-hybrid-651:job__NSDwtmHUGZX4UoA6l3-2qKdjyo

What steps will reproduce the problem?
1. Select extract table from the UI
2. Select CSV as the output format
3. Use a wildcard in the google cloud storage file name

```

Original issue reported on code.google.com by `bjohn...@musiccitydata.com` on 11 Aug 2014 at 9:16

**Comment by GoogleCodeExporter** [comment]
```
This case is pretty peculiar. I don't see any internal errors or warnings for 
your job--instead, we're taking way too long to read the input file and 
eventually timing out (3 times, at 2 hours each attempt). It looks like it's 
happening repeatedly, and only affecting your table.

I'll see if I can reproduce the problem tomorrow. In the meantime, you might 
try copying the table and then extracting the copy to see if that helps. (If 
you don't mind, please leave the original table in place for another day or so, 
so that I can investigate further.)

Thanks for the report, and apologies for the trouble!
```

Original comment by `jcon...@google.com` on 12 Aug 2014 at 6:24

**Comment by GoogleCodeExporter** [comment]
```
Thanks for the reply.

I was able to extract the data by breaking up the table into parts.  It still 
took way too long, but the smaller tables did download after running for 30 
minutes.

I copied the table and am running the extract on the entire copied table.  I 
will let you know what the result of that is.

Thanks, Brian
```

Original comment by `bjohn...@musiccitydata.com` on 12 Aug 2014 at 2:19

**Comment by GoogleCodeExporter** [comment]
```
I still get a backend error on the full table copy of the copy of the original 
table.  The issue seems to be around the sharding.

The last job that failed was savvy-hybrid-651:job_XzPha8qx4HF3GHDX1EHKDZwtftw.

Thanks, Brian
```

Original comment by `bjohn...@musiccitydata.com` on 12 Aug 2014 at 8:43

**Comment by GoogleCodeExporter** [comment]
```
Thanks--it definitely looks like it has something to do with the table itself 
(though not necessarily the specific sharding). Sorry for the delay--this is 
still on our list to investigate.
```

Original comment by `jcon...@google.com` on 13 Aug 2014 at 6:21

**Comment by GoogleCodeExporter** [comment]
Original comment by `roselan...@google.com` on 17 Aug 2014 at 6:41
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Quick question - is this still happening for you?  It's possibly/probably 
unrelated, but we were having some inter-datacentre network issues around the 
time that you opened this issue.
```

Original comment by `thomasp...@google.com` on 21 Aug 2014 at 9:19

**Comment by GoogleCodeExporter** [comment]
```
I am testing now.  The job for the test is: 

Job ID: savvy-hybrid-651:job_tfgP9xiG7Jc2KlO8HeRkRw5WvXQ
```

Original comment by `bjohn...@precisionprospecting.net` on 22 Aug 2014 at 2:59

**Comment by GoogleCodeExporter** [comment]
```
The issue does not seemed to be resolved.  The job has been running for over an 
hour and has only exported one part of the multi-part export file.
```

Original comment by `bjohn...@precisionprospecting.net` on 22 Aug 2014 at 4:18

**Comment by GoogleCodeExporter** [comment]
```
Hi Brian,

We are not experiencing any system-wide issues with export right now. I'm 
looking into what is causing this table to export slowly.

Thanks for your patience.
```

Original comment by `roselan...@google.com` on 22 Aug 2014 at 4:48

**Comment by GoogleCodeExporter** [comment]
```
The good news is the job is now completing.  The bad news is it took 2 hours to 
export 1.6GB.

I can work with this, but it would be nice if it ran faster.  

Thanks for investigating.  It is getting better.

Thanks,
Brian
```

Original comment by `bjohn...@precisionprospecting.net` on 22 Aug 2014 at 8:26

**Comment by GoogleCodeExporter** [comment]
```
Hi Brian,

I've looked into why your table takes so long to export. One factor may be that 
the table is highly compressed (less than 1/20th the size) in BigQuery and 
requires more processing power to decompress. To mitigate, we're upping 
resources for import/export jobs. Hopefully that makes things faster.

Another factor is that since the table is so small when compressed, it's just a 
few files in BigQuery, which limits the number of workers we can use to export 
it. However we're still not always optimal in choosing the number of workers. 
I'm looking into a fix for that.

I will update the issue when each of those fixes goes out, and we can see if 
that improves things.
```

Original comment by `roselan...@google.com` on 26 Aug 2014 at 3:20

**Comment by GoogleCodeExporter** [comment]
```
Hello.

I am still having this issue.  It is taking 2 hours to export a file that is 
225mb on google cloud storage.

We would like to start using big query in production, but the table extracts of 
the results are too slow.

JobID is imperial-glyph-494:bqjob_r69ce198f_000001486b5709db_1

This step used to take 5 minutes.  Please look into improving the extract 
process.

Brian
```

Original comment by `bjohn...@precisionprospecting.net` on 12 Sep 2014 at 8:45

**Comment by GoogleCodeExporter** [comment]
```
We'll investigate. Note that you can likely speed up the export by using a glob 
pattern (gs://foo/bar*.csv) which lets us run the export in parallel.
```

Original comment by `tig...@google.com` on 12 Sep 2014 at 11:38

**Comment by GoogleCodeExporter** [comment]
```
Hello.

I tried using the glob pattern.  The export still took 1 hour 12 minutes for 
250 mb worth of data.

The glob job was imperial-glyph-494:bqjob_r41af9497_000001487b4988a8_1

Something must have changed with the extract jobs.  They are much slower than 
they were six months ago.
```

Original comment by `bjohn...@musiccitydata.com` on 16 Sep 2014 at 12:45

**Comment by GoogleCodeExporter** [comment]
```
It appears that there was a regression for a certain class of tables, and we're 
currently working to identify the cause. Sorry for the inconvenience.
```

Original comment by `roselan...@google.com` on 17 Sep 2014 at 8:58

**Comment by GoogleCodeExporter** [comment]
```
We've observed that in some cases tables that take an egregiously long time to 
extract in CSV can be quite quick when either GZIP compression is enabled, or 
JSON is selected as the file format.

Please try that and let me know if it helps. Root cause investigation continues.
```

Original comment by `roselan...@google.com` on 24 Sep 2014 at 4:43

**Comment by GoogleCodeExporter** [comment]
```
Thanks for the update.  We also figured out the gzip trick through some 
experimentation on our end.  We will use that methodology going forward.  
Thanks for sending that tip along.

The issue with using gzip is we had to rewrite the rest of the code to 
accommodate.  Not a huge deal, but I am worried about the reliability of big 
query if a basic thing like an extract statement all of a sudden takes 20x as 
long to run.

Brian
```

Original comment by `bjohn...@precisionprospecting.net` on 24 Sep 2014 at 5:47

**Comment by GoogleCodeExporter** [comment]
```
It looks like BQ Extract just went down.

Please see job number: 
savvy-hybrid-651:bqjob_r75e29199779d84ec_00000148a96e09fc_1

I get the following error message when I try to export the table: 

BigQuery error in extract operation: Error processing job
'savvy-hybrid-651:bqjob_r75e29199779d84ec_00000148a96e09fc_1':
/bigstore/axial-analysis/hip_episode_claim_QA.csv

This export was working yesterday.  I did add 3 columns to the table and then 
reexport for analysis in python.

Thanks,
Brian
```

Original comment by `bjohn...@precisionprospecting.net` on 24 Sep 2014 at 8:54

**Comment by GoogleCodeExporter** [comment]
```
Regarding update #18: This was an separate issue that affected some export jobs 
between 13:10 and 15:25 PDT on 2014-09-24.  The issue was related to writing 
data to cloud storage during this interval.
```

Original comment by `sholly...@google.com` on 25 Sep 2014 at 4:58

**Comment by GoogleCodeExporter** [comment]
```
We've pushed a fix to address the long extract to CSV. Please retry your jobs 
with CSV extract and let me know if performance is still degraded.
```

Original comment by `roselan...@google.com` on 6 Oct 2014 at 11:43

**Comment by GoogleCodeExporter** [comment]
```
Thanks!

I should be able to test this tomorrow.

Brian
```

Original comment by `bjohn...@musiccitydata.com` on 7 Oct 2014 at 6:37

**Comment by GoogleCodeExporter** [comment]
```
YESS!!!!

Perfect.  It is working exactly as it was before.

Thanks for putting in the fix.  I feel much better about starting to move more 
of our production work over to Big Query now.

Thanks.  This was a huge deal for me.

Brian
```

Original comment by `bjohn...@musiccitydata.com` on 8 Oct 2014 at 5:19

**Comment by GoogleCodeExporter** [comment]
```
I'm glad to hear it! Don't hesitate to let us know if you run into any more 
problems.
```

Original comment by `roselan...@google.com` on 12 Oct 2014 at 12:42
- Changed state: **Verified**
